src/hotspot/cpu/s390/macroAssembler_s390.cpp

```diff
@@ -3134,18 +3109,150 @@ void MacroAssembler::check_klass_subtype_slow_path(Register Rsubklass,
+// lookup_secondary_supers_table() to do the work. Any of the temp
+// regs may be noreg, in which case this logic will choose some
+// registers push and pop them from the stack.
+void MacroAssembler::check_klass_subtype_slow_path_table(Register sub_klass,
+                                                         Register super_klass,
+                                                         Register temp_reg,
+                                                         Register temp2_reg,
+                                                         Register temp3_reg,
+                                                         Register temp4_reg,
+                                                         Register result_reg,
+                                                         Label* L_success,
+                                                         Label* L_failure,
+                                                         bool set_cond_codes) {
+  BLOCK_COMMENT("check_klass_subtype_slow_path_table {");
+
+  RegSet temps = RegSet::of(temp_reg, temp2_reg, temp3_reg, temp4_reg);
+
+  assert_different_registers(sub_klass, super_klass, temp_reg, temp2_reg, temp4_reg);
+
+  Label L_fallthrough;
+  int label_nulls = 0;
+  if (L_success == nullptr)   { L_success   = &L_fallthrough; label_nulls++; }
+  if (L_failure == nullptr)   { L_failure   = &L_fallthrough; label_nulls++; }
+  assert(label_nulls <= 1, "at most one null in the batch");
+
+  RegSetIterator<Register> available_regs
+  // Z_R0 will be used to hold Z_R15(Z_SP) while pushing a new frame, So don't use that here.
+  // Z_R1 is will be used to hold r_bitmap in lookup_secondary_supers_table_var, so can't be used
+  // Z_R2, Z_R3, Z_R4 will be used in secondary_supers_verify, for the failure reporting
+    = (RegSet::range(Z_R0, Z_R15) - temps - sub_klass - super_klass - Z_R1_scratch - Z_R0_scratch - Z_R2 - Z_R3 - Z_R4).begin();
+
+  RegSet pushed_regs;
+
+  temp_reg  = allocate_if_noreg(temp_reg,  available_regs, pushed_regs);
+  temp2_reg = allocate_if_noreg(temp2_reg, available_regs, pushed_regs);
+  temp3_reg = allocate_if_noreg(temp3_reg, available_regs, pushed_regs);;
+  temp4_reg = allocate_if_noreg(temp4_reg, available_regs, pushed_regs);
+  result_reg = allocate_if_noreg(result_reg, available_regs, pushed_regs);
+
+  const int frame_size = pushed_regs.size() * BytesPerWord + frame::z_abi_160_size;
+
+  // Push & save registers
+  {
+    int i = 0;
+    save_return_pc();
+    push_frame(frame_size);
+
+    for (auto it = pushed_regs.begin(); *it != noreg; i++) {
+      z_stg(*it++, i * BytesPerWord + frame::z_abi_160_size, Z_SP);
+    }
+    assert(i * BytesPerWord + frame::z_abi_160_size == frame_size, "sanity");
+  }
+
+  lookup_secondary_supers_table_var(sub_klass,
+                                    super_klass,
+                                    temp_reg, temp2_reg, temp3_reg, temp4_reg, result_reg);
+
+  z_cghi(result_reg, 0);
```

Comment:
> There is a significant (and varying) distance between here and where the CC is evaluated. I would expect at least a comment that the code in between must not alter the CC.

I'd just move the comparison nearer the conditional branch.

Reply:
>I'd just move the comparison nearer the conditional branch. 

result_reg could be `noreg`. So it will be popped out and we might loose the content. One such call is at `3245`:
```cpp
  if (UseSecondarySupersTable) {
    check_klass_subtype_slow_path_table(sub_klass,
                                        super_klass,
                                        temp_reg,
                                        temp2_reg,
                                        /*temp3*/noreg,
                                        /*temp4*/noreg,
                                        /*result*/noreg,
                                        L_success,
                                        L_failure,
                                        set_cond_codes);
```

But we can store the value in scratch register, maybe `Z_R0` and then move this check.

Reply:
> Also, between this instruction and the branch instruction, we are emitting `z_lg` and `z_lghi` instruction. So maybe just put a comment ?

Yes, do that. I notice we do the same on AArch64, but I sort-of "know" that `macroAssembler::pop` preserves flags on aarch64 and x86.

Reply:
Oh, that wasn't even my worse suggestion 😅. 

Another solution which I was thinking about was, passing `temp_reg` at place of `result_reg` then we can move that check down without any issue, because value wouldn't be altered. And `temp_reg` then will be allocated a new register from `RegSet`.